opencog/reasoning/RuleEngine/rule-engine-src/pln/BackwardChainer.cc

```diff
@@ -0,0 +1,492 @@
+			for (Handle h : premises) {
+				logical_link_premise_map[llink].push_back(h);
+				//check if h is a logical link type and push it for building next iter map
+				t = as_->getType(h);
+				it = find(logical_link_types_.begin(),
+						logical_link_types_.end(), t);
+				if (it != logical_link_types_.end())
+					logical_links.push_back(h);
+
+			}
+		} while (not logical_links.empty());
+		return logical_link_premise_map;
+	} else
+		return map<Handle, HandleSeq> { { Handle::UNDEFINED, HandleSeq {
+				root_llink } } };
+}
+HandleSeq BackwardChainer::chase_var_values(Handle& hvar,
+		vector<map<Handle, HandleSeq>>& inference_list, HandleSeq& results) {
+	for (auto it = inference_list.begin(); it != inference_list.end(); ++it) {
+		map<Handle, HandleSeq> var_value = *it;
+		if (var_value.count(hvar) != 0) {
+			HandleSeq values = var_value[hvar];
+			for (Handle h : values) {
+				if (as_->getType(h) == VARIABLE_NODE) {
+					HandleSeq new_values;
+					HandleSeq hseq = chase_var_values(h, inference_list,
+							new_values);
+					results.insert(results.end(), hseq.begin(), hseq.end());
+				} else {
+					results.push_back(h);
+				}
+			}
+		}
+	}
+	return results;
+}
+map<Handle, HandleSeq> BackwardChainer::ground_target_vars(Handle& hgoal,
+		vector<map<Handle, HandleSeq>>& inference_list) {
+	map<Handle, HandleSeq> vg_map;
+	HandleSeq hgoal_vars = commons_->get_nodes(hgoal, vector<Type> {
+			VARIABLE_NODE });
+
+	for (map<Handle, HandleSeq> vgm : inference_list) {
+		for (auto it = vgm.begin(); it != vgm.end(); ++it) {
+
+			Handle hvar = it->first;
+			auto i = find(hgoal_vars.begin(), hgoal_vars.end(), hvar);
+
+			if (i != hgoal_vars.end()) {
+				HandleSeq values;
+				HandleSeq groundings = it->second;
+				for (Handle h : groundings) {
+					if (as_->getType(h) == VARIABLE_NODE) {
+						HandleSeq val;
+						val = chase_var_values(h, inference_list, val);
+						values.insert(values.end(), val.begin(), val.end());
+					} else {
+						values.push_back(h);
+					}
+					//erase duplicates i.e union
+					sort(values.begin(), values.end());
+					values.erase(unique(values.begin(), values.end()),
+							values.end());
+				}
+				if (vg_map.count(hvar) == 0)
+					vg_map[hvar] = values;
+				else {
+					HandleSeq existing = vg_map[hvar];
+					sort(existing.begin(), existing.end());
+					sort(values.begin(), values.end());
+					HandleSeq hs_combined;
+					set_union(existing.begin(), existing.end(), values.begin(),
+							values.end(), back_inserter(hs_combined));
+					vg_map[hvar] = hs_combined;
+				}
+			}
+		}
+	}
+	return vg_map;
+}
+map<Handle, HandleSeq> BackwardChainer::unify_to_empty_set(Handle&htarget) {
+	HandleSeq vars = commons_->get_nodes(htarget,
+			vector<Type> { VARIABLE_NODE });
+	map<Handle, HandleSeq> result;
+	for (Handle h : vars)
+		result[h] = HandleSeq { Handle::UNDEFINED };
+	return result;
+}
+
+map<Handle, HandleSeq> BackwardChainer::unify(Handle& htarget, Handle& match,
```

Comment:
The name of  this method, and the last few methods, is confusing to me.  The pattern matcher is a unifier.  That is what it does: it unifies. Why do you need to create a new, different kind of unifier?